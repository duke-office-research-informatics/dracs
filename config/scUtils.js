import React from 'react';
import { ThemeProvider } from 'styled-components';
import { mount, shallow } from 'enzyme';
import { theme } from '../lib/dracs.es.js';

export const shallowWithTheme = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return shallow(tree, { context });
};

export const mountWithTheme = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();

  return mount(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes // Needed so child components receive the theme prop
  });
};

//FOR ACCESSING PORTALS VIA ENZYME -- import portal from dracs
// https://github.com/airbnb/enzyme/issues/1202#issuecomment-351639783
// function mountPortal(el){
//   const wrapper =  mountWithTheme(el);
//   wrapper.setState({ visible: true});
//   const portalInstance = wrapper.find(Portal).instance();
//   return mountWithTheme(portalInstance.props.children);
// }

//FOR ACCESSING EVENT LISTENERS ATTACHED TO DOCUMENT OR DOCUMENT.BODY
//https://github.com/airbnb/enzyme/issues/426#issuecomment-296304015
// const map = {};
// document.body.addEventListener = jest.fn((event, cb) => {
//   map[event] = cb;
// });
