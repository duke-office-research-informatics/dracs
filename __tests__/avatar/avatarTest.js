import React from "react";
import { mountWithTheme } from "../../config/scUtils.js";
import { axe } from "jest-axe";
import { theme, Avatar, IconProfile } from "../../lib/dracs.es.js";

const imgUrl =
  "https://image.freepik.com/free-vector/soft-background-with-a-cute-blue-watercolor-stain_1048-5169.jpg";
const dispLetter = "a";

describe("Avatar", () => {
  it("matches snapshot", () => {
    const wrapper = mountWithTheme(<Avatar />);
    expect(wrapper).toMatchSnapshot();
  });
  it("returns no a11y violations", async () => {
    const wrapper = mountWithTheme(<Avatar />);
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });

  describe("without props", () => {
    it("has a wrapper and the fallback profile icon", () => {
      const wrapper = mountWithTheme(<Avatar />);
      expect(wrapper.find("avatar__AvatarWrap").length).toBe(1);
      expect(wrapper.find(IconProfile).length).toBe(1);
    });
    it("wrapper node has correct styles", () => {
      const wrapper = mountWithTheme(<Avatar />);
      const div = wrapper.find("avatar__AvatarWrap");
      expect(div).toHaveStyleRule("display", "flex");
      expect(div).toHaveStyleRule("align-items", "center");
      expect(div).toHaveStyleRule("justify-content", "center");
      expect(div).toHaveStyleRule("width", "32px");
      expect(div).toHaveStyleRule("height", "32px");
      expect(div).toHaveStyleRule("min-width", "32px");
      expect(div).toHaveStyleRule("min-height", "32px");
      expect(div).toHaveStyleRule("background-repeat", "no-repeat");
      expect(div).toHaveStyleRule("background-size", "cover");
      expect(div).toHaveStyleRule("background-color", "transparent");
      expect(div).toHaveStyleRule("border-radius", "50%");
      expect(div).toHaveStyleRule("color", "transparent");
      expect(div).toHaveStyleRule("font-size", "0");
      expect(div).toHaveStyleRule("overflow", "hidden");
    });
    it("profile icon has correct styling", () => {
      const wrapper = mountWithTheme(<Avatar />);
      const svg = wrapper.find(IconProfile);
      expect(svg).toHaveStyleRule("fill", "#fff");
      expect(svg).toHaveStyleRule("vertical-align", "top");
    });
  });

  describe("`bgColor` prop", () => {
    it("matches snapshot", () => {
      const wrapper = mountWithTheme(<Avatar bgColor={theme.colors.muted} />);
      expect(wrapper).toMatchSnapshot();
    });
    it("applies the correct styling", () => {
      const wrapper = mountWithTheme(<Avatar bgColor={theme.colors.muted} />);
      const div = wrapper.find("avatar__AvatarWrap");
      expect(div).toHaveStyleRule("background-color", theme.colors.muted);
    });
  });

  describe("`displayLetter` prop", () => {
    it("matches snapshot", () => {
      const wrapper = mountWithTheme(<Avatar displayLetter={dispLetter} />);
      expect(wrapper).toMatchSnapshot();
    });
    it("returns no a11y violations", async () => {
      const wrapper = mountWithTheme(<Avatar displayLetter={dispLetter} />);
      const results = await axe(wrapper.html());
      expect(results).toHaveNoViolations();
    });
    it("applies the proper styles to the wrapper", () => {
      const wrapper = mountWithTheme(<Avatar displayLetter={dispLetter} />);
      const div = wrapper.find("avatar__AvatarWrap");
      expect(div).toHaveStyleRule("text-align", "center");
    });
    it("inserts the letter into the dom instead of the profile icon", () => {
      const wrapper = mountWithTheme(<Avatar displayLetter={dispLetter} />);
      const letter = wrapper.find("avatar__AvatarLetter");
      const icon = wrapper.find(IconProfile);
      expect(icon.length).toBe(0);
      expect(letter.length).toBe(1);
      expect(letter.text()).toBe(dispLetter);
    });
    it("applies correct styling to the letter", () => {
      const wrapper = mountWithTheme(<Avatar displayLetter={dispLetter} />);
      const letter = wrapper.find("avatar__AvatarLetter");
      expect(letter).toHaveStyleRule("position", "relative");
      expect(letter).toHaveStyleRule("top", `${(32 / 3.65).toFixed(3)}px`);
      expect(letter).toHaveStyleRule("color", "#fff");
      expect(letter).toHaveStyleRule("vertical-align", "top");
      expect(letter).toHaveStyleRule("font-size", `${32 / 1.5}px`);
      expect(letter).toHaveStyleRule("line-height", "0.8");
      expect(letter).toHaveStyleRule("text-transform", "uppercase");
    });
  });

  describe("`iconColor`", () => {
    it("matches snapshot", () => {
      const wrapper = mountWithTheme(
        <Avatar displayLetter={dispLetter} iconColor={theme.colors.action} />
      );
      expect(wrapper).toMatchSnapshot();
    });
    it("updates the default profile icon color", () => {
      const wrapper = mountWithTheme(
        <Avatar iconColor={theme.colors.action} />
      );
      const svg = wrapper.find(IconProfile);
      expect(svg).toHaveStyleRule("fill", theme.colors.action);
    });
    it("updates letter color style", () => {
      const wrapper = mountWithTheme(
        <Avatar displayLetter={dispLetter} iconColor={theme.colors.action} />
      );
      const letter = wrapper.find("avatar__AvatarLetter");
      expect(letter).toHaveStyleRule("color", theme.colors.action);
    });
  });

  describe("`iconHoverColor`", () => {
    it("matches snapshot", () => {
      const wrapper = mountWithTheme(
        <Avatar iconHoverColor={theme.colors.action} />
      );
      expect(wrapper).toMatchSnapshot();
    });
    it("updates the profile icon svg hover style", () => {
      const wrapper = mountWithTheme(
        <Avatar iconHoverColor={theme.colors.action} />
      );
      const svg = wrapper.find(IconProfile);
      expect(svg).toHaveStyleRule("fill", theme.colors.action, {
        modifier: ":hover",
      });
    });
  });

  describe("`imgUrl`", () => {
    it("matches snapshot", () => {
      const wrapper = mountWithTheme(<Avatar imgUrl={imgUrl} />);
      expect(wrapper).toMatchSnapshot();
    });
    it("returns no a11y violations", async () => {
      const wrapper = mountWithTheme(<Avatar imgUrl={imgUrl} />);
      const results = await axe(wrapper.html());
      expect(results).toHaveNoViolations();
    });
    it("does not display the profile icon or the letter", () => {
      const wrapper = mountWithTheme(<Avatar imgUrl={imgUrl} />);
      expect(wrapper.find(IconProfile).length).toBe(0);
      expect(wrapper.find("avatar__AvatarLetter").length).toBe(0);
    });
    it("applies the img and correct styles to the wrapper div", () => {
      const wrapper = mountWithTheme(<Avatar imgUrl={imgUrl} />);
      const div = wrapper.find("avatar__AvatarWrap");
      expect(div).toHaveStyleRule("display", "flex");
      expect(div).toHaveStyleRule("align-items", "center");
      expect(div).toHaveStyleRule("justify-content", "center");
      expect(div).toHaveStyleRule("background-image", `url(${imgUrl})`);
    });
  });

  describe("`size` (larger than 32)", () => {
    it("matches snapshot", () => {
      const wrapper = mountWithTheme(
        <Avatar displayLetter={dispLetter} size={40} />
      );
      expect(wrapper).toMatchSnapshot();
    });
    it("applies the correct styling to the letter and wrapper", () => {
      const wrapper = mountWithTheme(
        <Avatar displayLetter={dispLetter} size={40} />
      );
      const div = wrapper.find("avatar__AvatarWrap");
      const letter = wrapper.find("avatar__AvatarLetter");
      expect(div).toHaveStyleRule("width", "40px");
      expect(div).toHaveStyleRule("height", "40px");
      expect(div).toHaveStyleRule("min-width", "40px");
      expect(div).toHaveStyleRule("min-height", "40px");
      expect(letter).toHaveStyleRule("top", `${(40 / 4).toFixed(3)}px`);
      expect(letter).toHaveStyleRule("font-size", `${40 / 1.5}px`);
    });
  });
});
