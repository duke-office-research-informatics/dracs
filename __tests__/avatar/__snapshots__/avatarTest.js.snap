// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar \`bgColor\` prop matches snapshot 1`] = `
.c1 {
  fill: #fff;
  vertical-align: top;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  min-width: 32px;
  min-height: 32px;
  background-repeat: no-repeat;
  background-size: cover;
  background-color: #696969;
  border-radius: 50%;
  color: transparent;
  font-size: 0;
  overflow: hidden;
}

<div
  className="c0"
  size={32}
>
  <svg
    aria-label="profile icon"
    className="c1"
    color="#fff"
    height={30}
    role="img"
    viewBox="0 0 50 50"
    width={30}
  >
    <title>
      profile icon
    </title>
    <path
      d="M34.8,33.8c-1.7-0.7-3.8-1.6-3.8-2.3V27c2.5-1.9,4-4.9,4-8v-6c0-5.5-4.5-10-10-10c-5.5,0-10,4.5-10,10v6c0,3.1,1.5,6.2,4,8  v4.5c0,0.6-2.1,1.5-3.8,2.3C11.1,35.5,5,38.1,5,45v1h40v-1C45,38.1,38.9,35.5,34.8,33.8z"
    />
  </svg>
</div>
`;

exports[`Avatar \`displayLetter\` prop matches snapshot 1`] = `
.c0 {
  text-align: center;
  width: 32px;
  height: 32px;
  min-width: 32px;
  min-height: 32px;
  background-repeat: no-repeat;
  background-size: cover;
  background-color: transparent;
  border-radius: 50%;
  color: transparent;
  font-size: 0;
  overflow: hidden;
}

.c1 {
  position: relative;
  top: 8.767px;
  color: #fff;
  vertical-align: top;
  font-size: 21.333333333333332px;
  line-height: 0.8;
  text-transform: uppercase;
}

<div
  className="c0"
  size={32}
>
  <span
    className="c1"
    size={32}
  >
    a
  </span>
</div>
`;

exports[`Avatar \`iconColor\` matches snapshot 1`] = `
.c0 {
  text-align: center;
  width: 32px;
  height: 32px;
  min-width: 32px;
  min-height: 32px;
  background-repeat: no-repeat;
  background-size: cover;
  background-color: transparent;
  border-radius: 50%;
  color: transparent;
  font-size: 0;
  overflow: hidden;
}

.c1 {
  position: relative;
  top: 8.767px;
  color: #0680cd;
  vertical-align: top;
  font-size: 21.333333333333332px;
  line-height: 0.8;
  text-transform: uppercase;
}

<div
  className="c0"
  size={32}
>
  <span
    className="c1"
    color="#0680cd"
    size={32}
  >
    a
  </span>
</div>
`;

exports[`Avatar \`iconHoverColor\` matches snapshot 1`] = `
.c1 {
  fill: #fff;
  vertical-align: top;
}

.c1:hover {
  fill: #0680cd;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  min-width: 32px;
  min-height: 32px;
  background-repeat: no-repeat;
  background-size: cover;
  background-color: transparent;
  border-radius: 50%;
  color: transparent;
  font-size: 0;
  overflow: hidden;
}

<div
  className="c0"
  size={32}
>
  <svg
    aria-label="profile icon"
    className="c1"
    color="#fff"
    height={30}
    role="img"
    viewBox="0 0 50 50"
    width={30}
  >
    <title>
      profile icon
    </title>
    <path
      d="M34.8,33.8c-1.7-0.7-3.8-1.6-3.8-2.3V27c2.5-1.9,4-4.9,4-8v-6c0-5.5-4.5-10-10-10c-5.5,0-10,4.5-10,10v6c0,3.1,1.5,6.2,4,8  v4.5c0,0.6-2.1,1.5-3.8,2.3C11.1,35.5,5,38.1,5,45v1h40v-1C45,38.1,38.9,35.5,34.8,33.8z"
    />
  </svg>
</div>
`;

exports[`Avatar \`imgUrl\` matches snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  min-width: 32px;
  min-height: 32px;
  background-repeat: no-repeat;
  background-size: cover;
  background-color: transparent;
  border-radius: 50%;
  color: transparent;
  font-size: 0;
  overflow: hidden;
  background-image: url(https://image.freepik.com/free-vector/soft-background-with-a-cute-blue-watercolor-stain_1048-5169.jpg);
}

<div
  className="c0"
  size={32}
/>
`;

exports[`Avatar \`size\` (larger than 32) matches snapshot 1`] = `
.c0 {
  text-align: center;
  width: 40px;
  height: 40px;
  min-width: 40px;
  min-height: 40px;
  background-repeat: no-repeat;
  background-size: cover;
  background-color: transparent;
  border-radius: 50%;
  color: transparent;
  font-size: 0;
  overflow: hidden;
}

.c1 {
  position: relative;
  top: 10.000px;
  color: #fff;
  vertical-align: top;
  font-size: 26.666666666666668px;
  line-height: 0.8;
  text-transform: uppercase;
}

<div
  className="c0"
  size={40}
>
  <span
    className="c1"
    size={40}
  >
    a
  </span>
</div>
`;

exports[`Avatar matches snapshot 1`] = `
.c1 {
  fill: #fff;
  vertical-align: top;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  min-width: 32px;
  min-height: 32px;
  background-repeat: no-repeat;
  background-size: cover;
  background-color: transparent;
  border-radius: 50%;
  color: transparent;
  font-size: 0;
  overflow: hidden;
}

<div
  className="c0"
  size={32}
>
  <svg
    aria-label="profile icon"
    className="c1"
    color="#fff"
    height={30}
    role="img"
    viewBox="0 0 50 50"
    width={30}
  >
    <title>
      profile icon
    </title>
    <path
      d="M34.8,33.8c-1.7-0.7-3.8-1.6-3.8-2.3V27c2.5-1.9,4-4.9,4-8v-6c0-5.5-4.5-10-10-10c-5.5,0-10,4.5-10,10v6c0,3.1,1.5,6.2,4,8  v4.5c0,0.6-2.1,1.5-3.8,2.3C11.1,35.5,5,38.1,5,45v1h40v-1C45,38.1,38.9,35.5,34.8,33.8z"
    />
  </svg>
</div>
`;
